modules/table/src/main/java/org/apache/ignite/internal/table/distributed/gc/MvGc.java

```diff
@@ -0,0 +1,249 @@
+            if (updatedLwm != newLwm) {
+                return;
+            }
+
+            executor.submit(() -> inBusyLock(this::initNewGcBusy));
+        });
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        shutdownAndAwaitTermination(executor, 10, TimeUnit.SECONDS);
+    }
+
+    private void initNewGcBusy() {
+        storageHandlerByPartitionId.keySet().forEach(this::scheduleGcForStorage);
+    }
+
+    private void scheduleGcForStorage(TablePartitionId tablePartitionId) {
+        executor.submit(() -> inBusyLock(() -> {
+            GcStorageHandler storageHandler = storageHandlerByPartitionId.compute(tablePartitionId, (id, gcStorageHandler) -> {
+                if (gcStorageHandler == null) {
+                    // Storage has been removed from garbage collection.
+                    return gcStorageHandler;
+                }
+
+                boolean casResult = gcStorageHandler.gcInProgressFuture.compareAndSet(null, new CompletableFuture<>());
+
+                assert casResult : tablePartitionId;
+
+                return gcStorageHandler;
+            });
+
+            if (storageHandler == null) {
+                // Storage has been removed from garbage collection.
+                return;
+            }
+
+            CompletableFuture<Void> future = storageHandler.gcInProgressFuture.get();
+
+            assert future != null : tablePartitionId;
+
+            try {
+                boolean scheduleGcForStorageAgain = true;
+
+                for (int i = 0; i < GC_BUTCH_SIZE && scheduleGcForStorageAgain; i++) {
+                    HybridTimestamp lowWatermark = lowWatermarkReference.get();
+
+                    assert lowWatermark != null : tablePartitionId;
+
+                    // If storage has been deleted or there is no garbage, then for now we will stop collecting garbage for this storage.
+                    if (!storageHandlerByPartitionId.containsKey(tablePartitionId)
+                            || !storageHandler.storageUpdateHandler.vacuum(lowWatermark)) {
+                        scheduleGcForStorageAgain = false;
+                    }
+                }
+
+                if (scheduleGcForStorageAgain) {
+                    scheduleGcForStorage(tablePartitionId);
+                }
+
+                future.complete(null);
+            } catch (Throwable t) {
+                future.completeExceptionally(t);
+            } finally {
+                boolean casResult = storageHandler.gcInProgressFuture.compareAndSet(future, null);
+
+                assert casResult : tablePartitionId;
+            }
+        }));
+    }
+
+    private <T> T inBusyLock(Supplier<T> supplier) {
+        if (!busyLock.enterBusy()) {
+            throw new IgniteInternalException(Gc.CLOSED_ERR);
```

Comment:
I think that the class `Gc` requires a better name

Reply:
Try to rename